### PR TITLE
Add IG REC-track check

### LINKED ISSUE #1041: Use to ES8 await&async functions
## When to use await&async functions

```
// Promise function with requests
function requestW3cAPI(url) {
    return new Promise(resolve => {
        w3capi.fetch((err,data) =>{   // pseudo-code
            resolve(data);
        })
    });
}

// Specberus rule, call done() when finished
exports.check = async function () {
    var data = await requestW3cAPI(url);
    console.log(data); // will log result for api
    done();
}
check();
```

## ES8 Compatibility
Node14 which Specberus currently using supports ES8 well, so we can start using ES8 grammar.
https://node.green/#ES2017-features-async-functions-await

## What Need to Change
For Specberus, to pass jshint check, change `.jshintrc` config to `"esversion": 8`

https://github.com/w3c/specberus/blob/b06e394a2598aabb74e3a7f326c3e09effe170d4/.jshintrc#L2

### PR DESCRIPTION
**Pull request by jennyliang220**
Closes #1041 #1015

### CODE REVIEW

File: lib/rules/sotd/pp.js [hunk 1]
```diff
@@ -81,6 +84,22 @@ function findPP (candidates, sr) {
...
+            if (err) {
+                reject();
+            } else {
+                resolve(data.name);
+            }
+        });
+    });
+}
+
+    // deliverIds.some(deliverId => {
```
Comment (nschonni) [inline lib/rules/sotd/pp.js:100]: Delete?

File: lib/rules/sotd/pp.js [hunk 2]
```diff
@@ -103,6 +122,12 @@ exports.check = function (sr, done) {
     ,     pp2017 = "https://www.w3.org/Consortium/Patent-Policy-20170801/"
     ,     patent = (sr.config.patentPolicy === "pp2020") ? pp2020 : (sr.config.patentPolicy === "pp2004") ? pp2017: sr.config.patentPolicy
     ;
+    var groupName = await fetchGroupName(sr.getDelivererIDs());
+    sr.isIGDeliverable = groupName.endsWith('Interest Group');
+
+    // fetchGroupName(sr.getDelivererIDs()).then(function (groupName) {
```
Comment (nschonni) [inline lib/rules/sotd/pp.js:128]: Delete?

File: lib/rules/sotd/pp.js (hunk 1 above)
Comment (jennyliang220) [reply lib/rules/sotd/pp.js:100]: Hi @nschonni , thanks for the review. This PR is not finished, just push to view diff and discuss. This `deliverIds` is an array and need iteration, I'll continue working on it today.

File: lib/rules/sotd/pp.js (hunk 2 above)
Comment (jennyliang220) [reply lib/rules/sotd/pp.js:128]: Same reason I leave comment here, to discuss whether use `await` or Promise `then`. Now I tend to use `await` to make the code easy to read.

File: lib/rules/sotd/pp.js (hunk 1 above)
Comment (nschonni) [reply lib/rules/sotd/pp.js:100]: Sorry about that, I missed the Draft status

**Comment by jennyliang220** [comment]
Tried with `Promise.any` locally, although document is in [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any), still get `TypeError: Promise.any is not a function` when running. [Browser_compatibility](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any#Browser_compatibility) shows it's not available until Node15 (currently specberus uses Node14)
image

**Comment by nschonni** [comment]
Might be worth adding https://www.npmjs.com/package/eslint-plugin-promise

File: lib/rules/sotd/pp.js [hunk 3]
```diff
@@ -110,6 +130,15 @@ exports.check = function (sr, done) {
             return done();
         }
 
+        // Use Promise.all to get Group Names, if groups include IG, will check `noRec` sentence in patent policy paragraph.
+        await Promise.all(fetchGroupNames(sr.getDelivererIDs())).then(function (groupNames) {
+            sr.isIGDeliverable = groupNames.some(groupName => {
+                return groupName.endsWith('Interest Group');
+            });
```
Comment (deniak) [inline lib/rules/sotd/pp.js:137]:
```suggestion
            sr.isIGDeliverable = groupNames.some(groupName => groupName.endsWith('Interest Group'));
```

**Review by deniak — CHANGES_REQUESTED** [review]
The rule looks good. However, I don't understand the test you updated. The test should check that a WD published by an IG triggers an error if the sentence 'The group(s) do/does not expect this document to become a W3C Recommendation'. is missing.

File: lib/rules/sotd/pp.js (hunk 3 above)
Comment (jennyliang220) [reply lib/rules/sotd/pp.js:137]: Thanks for the reminder. Update for `doc/wg-note2.html` is adding two delivererID, but it shouldn't be in this paragraph. Will move to another paragraph.

File: lib/rules/sotd/pp.js [hunk 4]
```diff
@@ -21,7 +24,7 @@ function buildWanted (groups, sr) {
 
         var noRec = ((groups.length == 2) ? "The groups do not " : "The group does not ") +
             "expect this document to become a W3C Recommendation\\. ?";
-        if (config.recTrackStatus && config.noRecTrack)
+        if ((config.recTrackStatus && config.noRecTrack )|| sr.isIGDeliverable)
```
Comment (deniak) [inline lib/rules/sotd/pp.js:27]: `isIGDeliverable` doesn't need to be global. It's used only in that file.

**Review by deniak — CHANGES_REQUESTED** [review]
I'm still not getting how the test reflects what the rules is supposed to check.

File: lib/rules/sotd/pp.js (hunk 4 above)
Comment (jennyliang220) [reply lib/rules/sotd/pp.js:27]: > I'm still not getting how the test reflects what the rules is supposed to check.

In 'test/docs/headers/wg-note2.html' I add a paragraph, saying the document is delivered by '40318', '109611' two groups. The second id '109611' is a **Interest Group**, that means the sentence `noRec` should be in PP paragraph.

The paragraph pp.js rule is looking for:
> /^This document was produced by a group operating under the( 15 September 2020| 1 August 2017)? W3C Patent Policy\. ?The group does not expect this document to become a W3C Recommendation\. ?$/

But in 'test/docs/headers/wg-note2.html' there is no such paragraph. So in 'test/rules.js' I add error "sotd.pp.no-pp" when checking 'wg-note2.html'.

Comment (deniak) [reply lib/rules/sotd/pp.js]: > But in 'test/docs/headers/wg-note2.html' there is no such paragraph. So in 'test/rules.js' I add error "sotd.pp.no-pp" when checking 'wg-note2.html'.

The sentence is actually in that document but in a separate paragraph. The current test is confusing. I believe wg-note2.html was create to make sure both sentences are in the same paragraph.
Now, it's also mixing what we expect from an IG so in the end, we don't really know what we are testing.

**Pull request closed without merging**